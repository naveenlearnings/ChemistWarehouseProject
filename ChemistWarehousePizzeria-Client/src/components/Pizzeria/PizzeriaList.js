import React, { useState, useEffect } from 'react';
import { useNavigate, Link } from "react-router-dom";
import Pizza from "./Pizza";
import styles from "./PizzeriaList.module.css";
import { v4 as uuidv4 } from 'uuid';
import { toast } from "react-toastify";

function PizzeriaList() {
  const [pizzerias, setPizzerias] = useState([]);
  const [orderList, setOrderList] = useState([]);
  const [totalPrice, setTotalPrice] = useState(0);

  const navigate = useNavigate();

  useEffect(() => {
    async function fetchData() {
      try {
        //Practically we will not harcode the complete URL, we will use a base url from some configuration component using
        // the environment varaiable.
        const response = await fetch('https://localhost:7199/api/Pizzerias', {
          method: 'GET',
          mode: "cors",
          headers: { "Content-Type": "application/json" },
        });

        if (!response.ok) {
          throw new Error('Failed to fetch data');
        }
        const pizzerias = await response.json();
        const pizzeriasWithExtraToppings = pizzerias.map(pizzeria => ({
          ...pizzeria,
          menu: {
            ...pizzeria.menu,
            pizzas: pizzeria.menu.pizzas.map(pizza => ({
              ...pizza,
              extraToppings: [{ name: "Cheese", isSelected: false }, { name: "Capsicum", isSelected: false }, { name: "Tomatoes", isSelected: false }, { name: "Olives", isSelected: false }]
            }))
          }
        }));

        setPizzerias(pizzeriasWithExtraToppings);

      } catch (error) {
        console.error(error)
      }
    }
    fetchData();
  }, []);

  const orderPizza = (pizza) => {
    orderList.push(pizza);
    alert("cart updated");
    toast.success("Cart Updated with new Pizza!", {
      position: toast.POSITION.TOP_RIGHT,
      autoClose: false
    });
    calculatePrice();
  }

  const calculatePrice = () => {
    let price = 0;
    orderList.forEach((pizza) => {
      price += pizza.price;
      if (pizza.extraToppings) {
        price += pizza.extraToppings.filter((topping) => topping.isSelected).length;
      }
    });

    setTotalPrice(price);
  };

  const toggleTopping = (pizzeria, pizza, topping) => {
    const updatedPizzerias = pizzerias.map((p) => {
      if (p.id === pizzeria.id) {
        const updatedPizzas = p.menu.pizzas.map((piz) => {
          if (piz.id === pizza.id) {
            const updatedToppings = piz.extraToppings.map((top) => {
              if (top.name === topping.name) {
                return { ...top, isSelected: !top.isSelected };
              }
              return top;
            });
            return { ...piz, extraToppings: updatedToppings };
          }
          return piz;
        });
        return { ...p, menu: { ...p.menu, pizzas: updatedPizzas } };
      }
      return p;
    });
    setPizzerias(updatedPizzerias);
  };

  const submitOrder = async () => {
    try {
      const pizzas = [];
      orderList.forEach((pizza) => {
        const pizzaDto = {
          Id: pizza.id,
          LocationId: pizza.locationId,
          Quentity: 1,
        }

        let toppings = [];

        if (pizza.extraToppings) {
          pizza.extraToppings.forEach((top) => {
            if (top.isSelected) {
              toppings.push(top.name);
            }
          });
        }

        if (toppings.length > 0) {
          pizzaDto.Toppings = toppings.join(",");
        }

        pizzas.push(pizzaDto);
      });

      const order = {
        Name: "Someone",
        Address: "7 Mckay St Broadview, 5083",
        Pizzas: pizzas,
      }

      //Practically we will not harcode the complete URL, we will use a base url from some configuration component using
      // the environment varaiable.
      const response = await fetch('https://localhost:7199/api/Orders', {
        method: 'POST',
        mode: "cors",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify(order),
      });

      if (!response.ok) {
        throw new Error('Sorry, Order failed. Try again later.');
      }

      navigate("/Order");

    } catch (error) {
      console.log(error);
    }
  };

  return (
    <div className="App">
    
      <main>
        {pizzerias.map(pizzeria => (
          <div key={pizzeria.id} className={`${styles.pizzaContainer} container`}>
            <div className=''>
              <div className='col-md'>
                <h4>{pizzeria.name}</h4>
                <h6>Location:{pizzeria.location.name}  <Link className='btn btn-link' to={`/pizzeria/${pizzeria.id}`}>Edit</Link> </h6>

              </div>

            </div>
            <div>
              {pizzeria.menu.pizzas.map(pizza => {
                return (
                  <Pizza key={uuidv4()} pizzeria={pizzeria} pizza={pizza} orderPizza={orderPizza} toggleTopping={toggleTopping} />
                )
              })}
            </div>
          </div>
        ))}

      </main>
      <header className="alert alert-primary" role="alert">
        <div className='container'>
          <div className='row'>
            <h5 className="right">Number of Pizzas Ordered: {orderList.length}</h5>
          </div>
          <div className='row'>
            <span> <h6>Total price: ${totalPrice}</h6> (including Tax)</span>
          </div>

          <div className='row'>
            <div className='col-md'>
              <button className='btn btn-primary btn-sm' onClick={submitOrder}>Submit Order</button>
            </div>
          </div>
        </div>
      </header>
    </div>
  );
}

export default PizzeriaList;
