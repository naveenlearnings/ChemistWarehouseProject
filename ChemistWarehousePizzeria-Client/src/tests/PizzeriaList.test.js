import React from 'react';
import { screen, render, waitFor } from '@testing-library/react';
import { BrowserRouter as Router, Route, Routes } from 'react-router-dom';
import PizzeriaList from "../components/pizzeria/pizzeriaList";
import { createServer } from "miragejs";
import userEvent from "@testing-library/user-event";

let pizzeriaList = null;

let server;
let setupServer = () => {
    server = createServer({
      routes() {
        //Practically we will not harcode the complete URL, we will use a base url from some configuration component using
        // the environment varaiable.
        this.get(`https://localhost:7199/api/Pizzerias`, () => {
          return pizzeriaList;
        });
    }});
};

let teardownServer = () => {
    pizzeriaList = null;
    server.shutdown();
  };
  
describe('PizzeriaList', () => {
    beforeEach(async () => {
        jest.resetModules();
        setupServer();
        pizzeriaList = [
            {
              "id": 1,
              "name": "Preston Pizzeria",
              "locationId": 1,
              "location": {
                "id": 1,
                "name": "Preston"
              },
              "menu": {
                "id": 1,
                "name": "Main Menu",
                "pizzeriaId": null,
                "pizzas": [
                  {
                    "id": 1,
                    "name": "Capricciosa",
                    "locationId": 1,
                    "toppings": "Cheese, Ham, Mushrooms, Olives",
                    "quentity": 0,
                    "price": 20.00
                  },
                  {
                    "id": 2,
                    "name": "Mexicana",
                    "locationId": 1,
                    "toppings": "Cheese, Salami, Capsicum, Chilli",
                    "quentity": 0,
                    "price": 18.00
                  },
                  {
                    "id": 3,
                    "name": "Margherita",
                    "locationId": 1,
                    "toppings": "Cheese, Spinach, Ricotta, Cherry Tomatoes",
                    "quentity": 0,
                    "price": 22.00
                  }
                ]
              }
            },
            {
              "id": 2,
              "name": "Southbank Pizzeria",
              "locationId": 2,
              "location": {
                "id": 2,
                "name": "Southbank"
              },
              "menu": {
                "id": 2,
                "name": "Main Menu",
                "pizzeriaId": null,
                "pizzas": [
                  {
                    "id": 1,
                    "name": "Capricciosa",
                    "locationId": 2,
                    "toppings": "Cheese, Ham, Mushrooms, Olives",
                    "quentity": 0,
                    "price": 23.00
                  },
                  {
                    "id": 4,
                    "name": "Vegetarian1",
                    "locationId": 2,
                    "toppings": "Cheese, Mushrooms, Capsicum, Onion, Olives",
                    "quentity": 0,
                    "price": 17.00
                  }
                ]
              }
            }
          ];
      });
    
      afterEach(() => {
        teardownServer();
      });

    it('renders correctly', async () => {
        const pizzeriaList = render(
        <Router>
            <PizzeriaList />
        </Router>
        );
        expect(pizzeriaList).toMatchSnapshot();
    });

    it("calculates correct price", async ()=> {
        render(
            <Router>
                <PizzeriaList/>
            </Router>
        );
        expect(screen.getByRole('heading', {  name: /number of pizzas ordered/i})).toBeInTheDocument();
        await waitFor(()=>{
            expect(screen.getByText(/Margherita/)).toBeInTheDocument();
            expect(screen.getByRole('heading', {  name: /total price: \$0/i})).toBeInTheDocument();
            const addBtns = screen.getAllByRole('button', {  name: /add to order/i});
            userEvent.click(addBtns[0]);
            expect(screen.getByRole('heading', {  name: /total price: \$20/i})).toBeInTheDocument();
        });
    });
});
