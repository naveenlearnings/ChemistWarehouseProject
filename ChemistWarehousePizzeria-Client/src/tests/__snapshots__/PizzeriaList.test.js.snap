// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PizzeriaList renders correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="App"
      >
        <header
          class="alert alert-primary"
          role="alert"
        >
          <div
            class="container"
          >
            <div
              class="row"
            >
              <h1
                class="center"
              >
                Number of Pizzas Ordered: 
                0
              </h1>
            </div>
            <div
              class="row"
            >
              <h3>
                Total price: $
                0
              </h3>
            </div>
            <div
              class="row"
            >
              <div
                class="col-md"
              >
                <button
                  class="btn btn-primary"
                >
                  Submit Order
                </button>
              </div>
              <div
                class="col-md text-right"
              >
                <a
                  class="btn btn-primary mr-3 ml-auto"
                  href="/Pizzeria/new"
                >
                  Create New Pizzeria
                </a>
              </div>
            </div>
          </div>
        </header>
        <main />
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="App"
    >
      <header
        class="alert alert-primary"
        role="alert"
      >
        <div
          class="container"
        >
          <div
            class="row"
          >
            <h1
              class="center"
            >
              Number of Pizzas Ordered: 
              0
            </h1>
          </div>
          <div
            class="row"
          >
            <h3>
              Total price: $
              0
            </h3>
          </div>
          <div
            class="row"
          >
            <div
              class="col-md"
            >
              <button
                class="btn btn-primary"
              >
                Submit Order
              </button>
            </div>
            <div
              class="col-md text-right"
            >
              <a
                class="btn btn-primary mr-3 ml-auto"
                href="/Pizzeria/new"
              >
                Create New Pizzeria
              </a>
            </div>
          </div>
        </div>
      </header>
      <main />
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
